package nm.message;

message MessageWrapper {
	enum Type { CURSOR_MOVE = 1; SQUARE_FLAG = 2; SQUARE_OPEN = 3; PLAYER_JOIN = 4; PLAYER_QUIT = 5; WELCOME = 6; CHUNK_BYTES = 7; CHUNK_REQUEST = 8; CLEAR_AT = 9; }

	required Type type = 1;

	optional CursorMove cursorMove = 2;
	optional SquareOpen squareOpen = 3;
	optional ChunkBytes chunkBytes = 4;
	optional SquareFlag squareFlag = 5;
	optional Player playerJoin = 6;
	optional Player playerQuit = 7;
	optional Welcome welcome = 8;
	optional ChunkRequest chunkRequest = 9;
	optional ClearAt clearAt = 10;
}

message ClearAt {
	required int32 x = 1;
	required int32 y = 2;
}

message ChunkRequest {
	required int32 x = 1;
	required int32 y = 2;
}

message ChunkBytes {
	required int32 x = 1;
	required int32 y = 2;
	required bytes data = 3;
}

message CursorMove {
	required int32 x = 1;
	required int32 y = 2;
	optional int32 id = 3;
}

message SquareOpen {
	required int32 x = 1;
	required int32 y = 2;
	optional int32 id = 3;
}

message SquareFlag {
	required int32 x = 1;
	required int32 y = 2;
	optional int32 id = 3;
}

message Player {
	required int32 x = 1;
	required int32 y = 2;
	optional int32 id = 3;
}

message Welcome {
	required int32 version = 1;
	required int32 nPlayers = 2;
	repeated Player players = 3;
}
